### PR TITLE
Fix job selector validation and tests

### LINKED ISSUE #22920: Job created with no selector and  extra template labels fails validation
When creating a job like this:
```go
job := &extensions.Job{}
job.Name = "testjob"
job.Spec.Template.Labels = map[string]string{"foo": "bar"}
job.Spec.Template.Spec = ...
job.Spec.Template.Spec.RestartPolicy = kapi.RestartPolicyNever
_, err = client.Extensions().Jobs(testutil.Namespace()).Create(job)
```
the API returns the following error:
> Job.extensions "testjob" is invalid: spec.selector: Invalid value: {"matchLabels":{"controller-uid":"db5d1cc3-e8f0-11e5-914d-acbc32c1ca87","foo":"bar"}}: `selector` not auto-generated

This is the object passed to `ValidateGeneratedSelector`, which seems like it should validate:
```
(api.ObjectMeta) {
 Name: (string) (len=7) "testjob",
 GenerateName: (string) "",
 Namespace: (string) (len=11) "integration",
 SelfLink: (string) "",
 UID: (types.UID) (len=36) "db5d1cc3-e8f0-11e5-914d-acbc32c1ca87",
 ResourceVersion: (string) "",
 Generation: (int64) 0,
 CreationTimestamp: (unversioned.Time) 2016-03-13 03:54:47.477573961 -0400 EDT,
 DeletionTimestamp: (*unversioned.Time)(<nil>),
 DeletionGracePeriodSeconds: (*int64)(<nil>),
 Labels: (map[string]string) (len=1) {
  (string) (len=3) "foo": (string) (len=3) "bar"
 },
 Annotations: (map[string]string) <nil>
}
(extensions.JobSpec) {
 Parallelism: (*int)(0xc20b1caaa0)(1),
 Completions: (*int)(0xc20b1caaa8)(1),
 ActiveDeadlineSeconds: (*int64)(<nil>),
 Selector: (*unversioned.LabelSelector)(0xc209403220)({
  MatchLabels: (map[string]string) (len=2) {
   (string) (len=3) "foo": (string) (len=3) "bar",
   (string) (len=14) "controller-uid": (string) (len=36) "db5d1cc3-e8f0-11e5-914d-acbc32c1ca87"
  },
  MatchExpressions: ([]unversioned.LabelSelectorRequirement) <nil>
 }),
 ManualSelector: (*bool)(<nil>),
 Template: (api.PodTemplateSpec) {
  ObjectMeta: (api.ObjectMeta) {
   Name: (string) "",
   GenerateName: (string) "",
   Namespace: (string) "",
   SelfLink: (string) "",
   UID: (types.UID) "",
   ResourceVersion: (string) "",
   Generation: (int64) 0,
   CreationTimestamp: (unversioned.Time) 0001-01-01 00:00:00 +0000 UTC,
   DeletionTimestamp: (*unversioned.Time)(<nil>),
   DeletionGracePeriodSeconds: (*int64)(<nil>),
   Labels: (map[string]string) (len=3) {
    (string) (len=3) "foo": (string) (len=3) "bar",
    (string) (len=8) "job-name": (string) (len=7) "testjob",
    (string) (len=14) "controller-uid": (string) (len=36) "db5d1cc3-e8f0-11e5-914d-acbc32c1ca87"
   },
   Annotations: (map[string]string) <nil>
  },
  Spec: (api.PodSpec) {
   Volumes: ([]api.Volume) <nil>,
   Containers: ([]api.Container) (len=1 cap=1) {
    (api.Container) {
     Name: (string) (len=9) "container",
     Image: (string) (len=10) "test/image",
     Command: ([]string) <nil>,
     Args: ([]string) <nil>,
     WorkingDir: (string) "",
     Ports: ([]api.ContainerPort) <nil>,
     Env: ([]api.EnvVar) <nil>,
     Resources: (api.ResourceRequirements) {
      Limits: (api.ResourceList) <nil>,
      Requests: (api.ResourceList) <nil>
     },
     VolumeMounts: ([]api.VolumeMount) <nil>,
     LivenessProbe: (*api.Probe)(<nil>),
     ReadinessProbe: (*api.Probe)(<nil>),
     Lifecycle: (*api.Lifecycle)(<nil>),
     TerminationMessagePath: (string) (len=20) "/dev/termination-log",
     ImagePullPolicy: (api.PullPolicy) (len=6) "Always",
     SecurityContext: (*api.SecurityContext)(<nil>),
     Stdin: (bool) false,
     StdinOnce: (bool) false,
     TTY: (bool) false
    }
   },
   RestartPolicy: (api.RestartPolicy) (len=5) "Never",
   TerminationGracePeriodSeconds: (*int64)(0xc20b1caab8)(30),
   ActiveDeadlineSeconds: (*int64)(<nil>),
   DNSPolicy: (api.DNSPolicy) (len=12) "ClusterFirst",
   NodeSelector: (map[string]string) <nil>,
   ServiceAccountName: (string) "",
   NodeName: (string) (len=20) "nodename.example.com",
   SecurityContext: (*api.PodSecurityContext)(0xc20943c5c0)({
    HostNetwork: (bool) false,
    HostPID: (bool) false,
    HostIPC: (bool) false,
    SELinuxOptions: (*api.SELinuxOptions)(<nil>),
    RunAsUser: (*int64)(<nil>),
    RunAsNonRoot: (*bool)(<nil>),
    SupplementalGroups: ([]int64) <nil>,
    FSGroup: (*int64)(<nil>)
   }),
   ImagePullSecrets: ([]api.LocalObjectReference) <nil>
  }
 }
}
```

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #22920

### CODE REVIEW

File: pkg/apis/extensions/validation/validation.go [hunk 1]
```diff
@@ -419,6 +419,10 @@ func ValidateGeneratedSelector(obj *extensions.Job) field.ErrorList {
 	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "controller-uid", string(obj.UID))...)
 	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "job-name", string(obj.Name))...)
 	expectedLabels := make(map[string]string)
+	// Expect all the labels that are actually on the template
```
Comment (liggitt) [inline pkg/apis/extensions/validation/validation.go]: this allows a generated selector that has copied labels from the template to match those same custom labels in the template

File: pkg/apis/extensions/validation/validation_test.go [hunk 2]
```diff
@@ -1023,10 +1023,38 @@ func TestValidateJob(t *testing.T) {
 			},
 			Spec: extensions.JobSpec{
 				Selector:       validGeneratedSelector,
-				ManualSelector: newBool(true),
+				ManualSelector: newBool(false),
```
Comment (liggitt) [inline pkg/apis/extensions/validation/validation_test.go]: fixing a typo

File: pkg/apis/extensions/validation/validation_test.go [hunk 3]
```diff
@@ -980,7 +980,7 @@ func TestValidateJob(t *testing.T) {
 		MatchLabels: map[string]string{"a": "b"},
 	}
 	validGeneratedSelector := &unversioned.LabelSelector{
-		MatchLabels: map[string]string{"collection-uid": "1a2b3c"},
+		MatchLabels: map[string]string{"controller-uid": "1a2b3c", "job-name":"myjob"},
 	}
```
Comment (liggitt) [inline pkg/apis/extensions/validation/validation_test.go]: once this was validated as a generated selector, it required these labels

**Comment by liggitt** [comment]
@erictune PTAL

**Comment by ncdc** [comment]
cc @soltysh

**Comment by soltysh** [comment]
@liggitt you need to run `gofmt` on `pkg/apis/extensions/validation/validation_test.go`
Otherwise LGTM

@bgrant0607 @erictune if possible I'd still include this in 1.2 release.

**Comment by liggitt** [comment]
fixed

**Comment by liggitt** [comment]
marking as candidate, I don't have a good feel for the impact of the validation error, I just know I hit it

**Comment by bgrant0607** [comment]
Yes, we should get this into 1.2.

**Comment by liggitt** [comment]
Test failure looks to be https://github.com/kubernetes/kubernetes/issues/17998

**Comment by liggitt** [comment]
@k8s-bot test this issue #17998

**Comment by bgrant0607** [comment]
Hmm. Will comment on the issue.

**Comment by erictune** [comment]
I'm taking a look at this to see if it is safe.

However, the issue that motivated it, is, in my view, not a bug, but working as intended.  So, lowering priority to P2.  

Basically, all go clients of Job have to change for v1.2.

**Comment by erictune** [comment]
I don't think this rises to the level of cherry-pick at this late stage in the game, so I removed the cherrypick-candidate label.
I will mail kubernetes-dev and improve the release notes to make sure people know they need to edit their go client code.

**Comment by liggitt** [comment]
> Basically, all go clients of Job have to change for v1.2. 

I'm willing to do that, but I didn't understand what I was supposed to change to...

**Comment by erictune** [comment]
If you said `job.Spec.ManualSelector = true` then you would have gotten the old behavior.

**Comment by liggitt** [comment]
but I wanted an automatic selector... do I need to set manualSelector=false explicitly?

**Comment by liggitt** [comment]
just recreated the behavior directly against the API with `autoSelector:true` on a job with a pod template containing labels. See https://github.com/kubernetes/kubernetes/issues/22920#issuecomment-197140335

**Comment by bgrant0607** [comment]
Commented here: https://github.com/kubernetes/kubernetes/issues/22920#issuecomment-197158021

**Comment by liggitt** [comment]
Updated with change in defaulting.

**Comment by bgrant0607** [comment]
Thanks much.

Some of the defaults tests should actually fail validation, but this test doesn't run validation, so the exhaustive testing makes sense.

We should add some e2e tests that use autoSelector at some point.

LGTM

**Comment by erictune** [comment]
FYI test/e2e/batch_v1_jobs.go uses autoSelector

**Comment by bgrant0607** [comment]
@erictune Thanks for the pointer. v1beta1 with autoSelector was needed, I guess.

**Comment by erictune** [comment]
@liggitt thanks for this fix and the thorough testing.  This LGTM.  Sorry for dismissing your original report before reading closely enough to see what was going on.

**Pull request merged**